### PR TITLE
DROOLS-1679 revert and empty-list if any error in FEEL list filtering

### PR DESCRIPTION
**Pull request by tarilabs**
... when filter returns empty collection when there are errors
evaluating filter.

Continued from commit 6c8d3f9a1b6c3ba420445c99f50a65f066a01ca7
https://github.com/kiegroup/drools/pull/1514

@etirelli can you kindly check for me the additional tests are correct, in the sense they are indeed respecting the DMN v1.1 specification, please? Thx
(I am discovering these edge cases as a side-effect of implementing the compilation)

/cc @baldimir @kurobako @winklerm

### CODE REVIEW

**Review by baldimir — APPROVED** [review]
Looks good, however I didn't check the spec compliance. I am leaving this to @etirelli.

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/FEELListsTest.java [hunk 1]
```diff
@@ -70,6 +70,11 @@
 
                 // Other filtering
                 {"[\"a\", \"b\", \"c\"][a]", null, FEELEvent.Severity.ERROR },
+                {"{ a list : [ { a : false, b : 2 }, { a : true, b : 3 } ], r : a list[a] }.r", Arrays.asList( new HashMap<String, Object>() {{ put("a", true);  put("b", BigDecimal.valueOf( 3 )); } } ), null },
+                {"{ a list : [ \"a\", \"b\", \"c\" ], x : 2, a : a list[x]}.a", "b", null },
+                {"{ a list : [ { x : false, y : 2 }, { x : true, y : 3 } ], x : \"asd\", a : a list[x] }.a", Arrays.asList( new HashMap<String, Object>() {{ put("x", true);  put("y", BigDecimal.valueOf( 3 )); } } ), null },
```
Comment (tarilabs) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/FEELListsTest.java]:
```
{ a list : [ { x : false, y : 2 }, { x : true, y : 3 } ], x : \"asd\", a : a list[x] }
```

`a list` is a list
`x` is NOT a number
spec Table 54: Semantics of lists
page 127 of the printed PDF

> e1 is a list and e2 is an integer

first row is NOT this case
so
it might be row 3

> e1 is a list and type(FEEL(e2 , s')) is boolean

as long as when X in the special scope S' would exist, as a boolean
which it does
because the element of the list contains a definition for an attribute X
in the special scope S' : `{ x : false`

**Review by tarilabs — COMMENTED** [review]
summary of as-discussed on IRC with @baldimir

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/FEELListsTest.java [hunk 2]
```diff
@@ -70,6 +70,11 @@
...
+                {"{ a list : [ { x : false, y : 2 }, { x : true, y : 3 } ], x : false, a : a list[x] }.a", Arrays.asList( new HashMap<String, Object>() {{ put("x", true);  put("y", BigDecimal.valueOf( 3 )); } } ), null },
+                {"{ a list : [ { x : false, y : 2 }, { x : true, y : 3 } ], x : null, a : a list[x] }.a", Arrays.asList( new HashMap<String, Object>() {{ put("x", true);  put("y", BigDecimal.valueOf( 3 )); } } ), null },
```
Comment (etirelli) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/FEELListsTest.java]: @tarilabs all 5 test cases you are adding here make sense to me. An attribute in the relation will shadow any variable previously defined.

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/FilterExpressionNode.java [hunk 3]
```diff
@@ -94,27 +94,29 @@ public Object evaluate(EvaluationContext ctx) {
...
+                // element using the name item, unless the list element is a context that contains the key "item".
+                ctx.setValue("item", v);
+
+                // using Root object logic to avoid having to eagerly inspect all attributes.
+                ctx.setRootObject(v);
+
+                Object r = this.filter.evaluate(ctx);
+                if (r == null) {
+                    // I was expecting to handle "filter" as a Predicate, but is not a Predicate.
+                    return null;
```
Comment (etirelli) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/FilterExpressionNode.java]: I am not sure about this @tarilabs . 

An error when evaluating the filter for one element should not stop the evaluation. My understanding is that an error in this case will work as a "false" filter, and as such, simply skip the element. In other words, shouldn't this be "continue" instead of "return null"? Or, in case the error is on the predicate itself, not in the evaluation of a given element, then should be "break" instead of "return null".

I am pretty sure there is a ticket that says an error on a filter should return an empty list, not null, and we changed the behaviour in the past to do so.

Comment (etirelli) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/FilterExpressionNode.java]: Page 113 of the spec says: "The filter expression is evaluated for each item in list, and a list containing only items where the filter expression is true is returned."

Based on other similar cases in the spec, for a filter that returns any value other than "true", including "null" or errors, the result should be an empty list, not null.

Comment (tarilabs) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/FilterExpressionNode.java]: > I am pretty sure there is a ticket that says an error on a filter should return an empty list, not null, and we changed the behaviour in the past to do so.

That is indeed DROOLS-1679, which did not cover correctly for these additional cases, just line 72 in the test (just above the new introduced tests).

Comment (tarilabs) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/FilterExpressionNode.java]: > Based on other similar cases in the spec, for a filter that returns any value other than "true", including "null" or errors, the result should be an empty list, not null.

You might be right, the problem is we don't have yet test cases which cover this possibility (as the CI is green). I will write some tests to make sure this coverage of edge-cases is not exposing bugs in other scenarii.

Comment (tarilabs) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/FilterExpressionNode.java]: > Based on other similar cases in the spec, for a filter that returns any value other than "true", including "null" or errors, the result should be an empty list, not null.

@etirelli In other words we are saying original DROOLS-1679 was actually not a bug.

I tried to make a test case which is a bit more "edge":
```
{ a list : [ { a : false, b : 2 }, { a : null, b : 3 }, { b : 4 } ], r : a list[a] }.r
```
based on the spec, we are saying for
`a list[a]`
the filer: `a`
> is evaluated for each item in list, and a list containing only items where the filter expression is true is returned

so:
- when evaluated on the first element `{ a : false, b : 2 }` is false boolean-literal, hence not selected, 
- when evaluated on the second element `{ a : null, b : 3 }` is the null-literal, hence not selected, and 
- when evaluated on the third element `{ b : 4 }` it would be an error, hence null, hence not selected

But following this logic, the original test provided with DROOLS-1679:
```
["a", "b", "c"][a]
```
should return an empty list, not a null-literal and an error as per requested DROOLS-1679 itself.

What do you think @etirelli ?

Comment (etirelli) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/FilterExpressionNode.java]: @tarilabs I believe your analysis is correct, and so DROOLS-1679 was a mistake. We can bring this up today at the TCK meeting and hear some opinions.

Comment (etirelli) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/FilterExpressionNode.java]: Matteo, I discussed this further today with Bruce and he is actually having problems with this case when importing data from XML, as XML not always have all the attributes set. 

Can you go ahead and fix this as per our previous discussion? I.e., a filter would always return a list with all the elements for which the filter is true. In case any element fails in there or the filter expression returns null, it will only exclude the element, but will continue to process the list. In case all elements fail, the result will be an empty list.

Also, for cases like the 3rd bullet in your example, a missing attribute in a map should be considered null, not an error. 

So, in the following example:
```
{
    people : [ { firstName : "bob" }, { firstName : "max" } ]
    result : people[ lastName = null ]
}
```

result will be evaluated to:
```
 [ { firstName : "bob" }, { firstName : "max" } ]
```

Comment (tarilabs) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/FilterExpressionNode.java]: Implemented, and this test is also included.

**Comment by tarilabs** [comment]
I rebased the PR branch on the following:

> a filter would always return a list with all the elements for which the filter is true. In case any element fails in there or the filter expression returns null, it will only exclude the element, but will continue to process the list. In case all elements fail, the result will be an empty list. Also a missing attribute in a map [being the map an element in the list] should be considered null, not an error.

/cc @etirelli @baldimir @kurobako @winklerm

**Review by etirelli — APPROVED** [review]

**Comment by etirelli** [comment]
Looks good @tarilabs, thanks!
